### PR TITLE
frontend: Show feature flags in topbar and separate page

### PR DESCRIPTION
**Pull request by almusil**
As discussed in #1129. Opening the frontend part for discussion. 

Current implementation look:
Feature flags

### CODE REVIEW

**Comment by pietroalbini** [comment]
I'm not sure showing the additional dependencies is much useful, we could simplify the layout like this (showing the features on two columns on desktop):

photo_2020-10-30_11-41-39

**Comment by almusil** [comment]
> I'm not sure showing the additional dependencies is much useful, we could simplify the layout like this (showing the features on > two columns on desktop):

From my point of view it's useful to know which feature gets enabled by which flag. Especially for default.

**Comment by jyn514** [comment]
I also prefer having the features that are enabled shown.

**Comment by Kixiron** [comment]
I think we should show full detail, but on a separate page (placed alongside `/builds`) with the top bar linking to it

**Comment by Kixiron** [comment]
Made a very rough and out of proportion sketch of what I'm thinking, display the available features on the left, clicking on each "maximizes" it to show all the features and dependencies it activates. This allows us to give as much info as possible (ex. saying if the feature is actually a feature or just an optional dependency) without immediately overwhelming the user. 

image

**Comment by almusil** [comment]
Ok, I have been able to come up with this (I'm really bad at frontend, so please excuse some rough edges):
Screenshot_2020-11-05 regex 1 3 1 - Docs rs
Screenshot_2020-11-05 regex 1 3 1 - Docs rs(1)
Screenshot_2020-11-05 regex 1 3 1 - Docs rs(2)

**Comment by jyn514** [comment]
I like it! It could be nice to split it into columns like @pietroalbini  suggested, but I don't think we need to block on that.

File: templates/crate/features.html [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+                        <h3 id="{{ feature.name }}">{{ feature.name }}</h3>
+                        <ul class="pure-menu-list">
+                            {%- if feature.subfeatures -%}
+                                {%- for subfeature in feature.subfeatures -%}
+                                    <li class="pure-menu-item">
+                                        <span>{{ subfeature }}</span>
+                                    </li>
+                                {%- endfor -%}
+                            {%- else -%}
+                                <p>This feature flag do not enable additional features.</p>
```
Comment (jyn514) [inline templates/crate/features.html:60]:
```suggestion
                                <p>This feature flag does not enable additional features.</p>
```

File: src/db/types.rs [hunk 2]
```diff
@@ -12,4 +12,16 @@ impl Feature {
...
+        self.name.starts_with('_')
+    }
+
+    pub fn is_default(&self) -> bool {
+        self.name == "default"
+    }
+
+    pub fn subfeature_len(&self) -> usize {
+        self.subfeatures.len()
+    }
```
Comment (jyn514) [inline src/db/types.rs:26]: Is this function necessary? It seems like it would be more clear inlined.

File: src/web/crate_details.rs [hunk 3]
```diff
@@ -741,4 +744,42 @@ mod tests {
             Ok(())
         });
     }
+
+    #[test]
+    fn feature_flags_is_hidden_when_empty() {
```
Comment (jyn514) [inline src/web/crate_details.rs:749]:
```suggestion
    fn feature_flags_are_hidden_when_empty() {
```
or just
```suggestion
    fn feature_flags_hidden_when_empty() {
```

File: templates/crate/features.html [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+                            {%- for feature in features -%}
+                                <li class="pure-menu-item">
+                                    <a href="#{{ feature.name }}" class="pure-menu-link" style="text-align:center;">
+                                        {{ feature.name }}
+                                    </a>
+                                </li>
+                            {%- endfor -%}
+                        {%- else -%}
+                            <li class="pure-menu-item">
+                                <span style="font-size: 13px;">Feature flags are not available for this version.</span>
```
Comment (jyn514) [inline templates/crate/features.html:39]: It's possible they may not be available for any version.
```suggestion
                                <span style="font-size: 13px;">Feature flags are not available for this release.</span>
```

File: templates/rustdoc/topbar.html [hunk 5]
```diff
@@ -226,7 +226,27 @@
...
+
+            {# Build the dropdown list showing available features #}
+            <ul class="pure-menu-children">
+                {%- for feature in metadata.features -%}
+                {%- set target_url = "/crate/" ~ metadata.name ~ "/" ~ metadata.version ~ "/features#" ~ feature.name -%}
+                    <li class="pure-menu-item">
+                        <a href="{{ target_url | safe }}" class="pure-menu-link">{{ feature.name }}</a>
+                    </li>
+                {%- endfor -%}
+            </ul>
```
Comment (jyn514) [inline templates/rustdoc/topbar.html:247]: This gets a little overwhelming with a lot of features. Can you have just a single link to `/crate/.../features` instead, the way you have it on /crate?

image

File: templates/crate/features.html [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+                            </li>
+                        {%- endif -%}
+                    </ul>
+                </div>
+            </div>
+
+            <div class="pure-u-1 pure-u-sm-17-24 pure-u-md-19-24 package-details" id="main">
+                <h1>{{ metadata.name }}</h1>
+                {%- if features -%}
+                    <p>This version has <b>{{ feature_len }}</b> feature flags, <b>{{ default_feature_len }}</b> of them being enabled by <b>default</b>.</p>
```
Comment (jyn514) [inline templates/crate/features.html:49]:
```suggestion
                    <p>This version has <b>{{ features | length }}</b> feature flags, <b>{{ features[0] | length }}</b> of them being enabled by <b>default</b>.</p>
```
Then you can remove `feature_len` and `default_feature_len` from the template data. You may have to handle the case where `features | length` is 0 specially to avoid out of bounds errors.

File: src/web/crate_details.rs [hunk 7]
```diff
@@ -741,4 +744,42 @@ mod tests {
...
+        wrapper(|env| {
+            env.fake_release()
+                .name("binary")
+                .version("0.1.0")
+                .binary(true)
```
Comment (jyn514) [inline src/web/crate_details.rs:754]: Rather than use this hack, could you change `Fakes::features` to take an `Option`?

File: src/web/crate_details.rs [hunk 8]
```diff
@@ -741,4 +744,42 @@ mod tests {
...
+        wrapper(|env| {
+            let features = [("_private".into(), Vec::new())]
+                .iter()
+                .cloned()
+                .collect::<HashMap<String, Vec<String>>>();
+            env.fake_release()
+                .name("binary")
+                .version("0.1.0")
+                .binary(true)
+                .features(features)
```
Comment (jyn514) [inline src/web/crate_details.rs:776]: Because you hid the features with `binary(true)`, this doesn't test anything. Please change this to a library instead.

File: src/web/features.rs [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+             FROM builds
+             INNER JOIN releases ON releases.id = builds.rid
+             INNER JOIN crates ON releases.crate_id = crates.id
+             WHERE crates.name = $1 AND releases.version = $2
+             ORDER BY releases.id DESC",
+            &[&name, &version]
+        )
+    );
+    let row = cexpect!(req, query.get(0));
+    let features = MetaData::parse_features(row.get("features"));
```
Comment (jyn514) [inline src/web/features.rs:44]: This is already a field on `MetaData`, no need to parse and `SELECT` it twice. You can change everything using this to use `metadata.features` instead, then remove it from the SQL query.

Alternatively, since the /features page is the only one using this data, you can remove it from `MetaData` altogether and only have it for `/features`. But it definitely shouldn't be in both.

**Review by jyn514 — COMMENTED** [review]
Thanks for sticking with this :)

File: src/db/types.rs (hunk 2 above)
Comment (almusil) [reply src/db/types.rs:26]: To be honest I didn't know that you can do something like that, thanks

File: src/web/crate_details.rs (hunk 3 above)
Comment (almusil) [reply src/web/crate_details.rs:749]: Done

File: templates/crate/features.html (hunk 4 above)
Comment (almusil) [reply templates/crate/features.html:39]: Done

File: templates/rustdoc/topbar.html (hunk 5 above)
Comment (almusil) [reply templates/rustdoc/topbar.html:247]: Done (Hopefully I understood it well)

File: src/web/crate_details.rs (hunk 8 above)
Comment (almusil) [reply src/web/crate_details.rs:776]: Done

File: src/web/features.rs (hunk 9 above)
Comment (almusil) [reply src/web/features.rs:44]: Good idea, thanks

File: templates/crate/features.html (hunk 6 above)
Comment (almusil) [reply templates/crate/features.html]: Done

File: src/web/crate_details.rs (hunk 7 above)
Comment (almusil) [reply src/web/crate_details.rs]: Done

**Comment by jyn514** [comment]
@almusil it's not a big deal, but in the future it would be nice to add changes as additional commits instead of force-pushing. The diff github shows also includes the changes from master. Fortunately here there weren't many changes, but in a big repo like rust-lang/rust it would be unreadable.

File: src/web/crate_details.rs [hunk 10]
```diff
@@ -741,4 +744,79 @@ mod tests {
...
         });
     }
+
+    #[test]
+    fn feature_flags_report_empty() {
+        wrapper(|env| {
+            env.fake_release()
+                .name("library")
+                .version("0.1.0")
+                .binary(false)
```
Comment (jyn514) [inline src/web/crate_details.rs:754]: This is the default.
```suggestion
```

**Comment by almusil** [comment]
> @almusil it's not a big deal, but in the future it would be nice to add changes as additional commits instead of force-pushing. The diff github shows also includes the changes from master. Fortunately here there weren't many changes, but in a big repo like rust-lang/rust it would be unreadable.
> 
> (FWIW I consider this a bug in github [isaacs/github#1834](https://github.com/isaacs/github/issues/1834))

Ah right, the ugly thing about that is that every repo has different policy about this. I'll keep it in mind for future changes in this repo, thanks :)

File: src/test/fakes.rs [hunk 11]
```diff
@@ -210,6 +210,15 @@ impl<'a> FakeRelease<'a> {
         self
     }
 
+    pub(crate) fn features(mut self, opt_features: Option<HashMap<String, Vec<String>>>) -> Self {
+        if let Some(features) = opt_features {
+            self.package.features = features;
+        } else {
+            self.package.features = HashMap::new();
+        }
```
Comment (jyn514) [inline src/test/fakes.rs:218]: Wait, I'm confused - if the `features` field on package isn't optional, then how does this render pages for old docs? Does it confuse 'no features' with 'no data'? `[features]` with no entries is also valid, please add a test for that.

I ask because this definitely looks wrong - now there's two ways to say `HashMap::new()`, one by passing it in and one by passing in `None`.

File: templates/crate/features.html [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+            <div class="pure-u-1 pure-u-sm-17-24 pure-u-md-19-24 package-details" id="main">
+                <h1>{{ metadata.name }}</h1>
+                {%- if metadata.features -%}
+                    <p>This version has <b>{{ metadata.features | length }}</b> feature flags, <b data-id="default-feature-len">
+                        {%- if metadata.features[0].name == 'default' -%}
+                            {{ metadata.features[0].subfeatures | length }}
+                        {%- else -%}
+                            0
+                        {%- endif -%}
+                    </b> of them being enabled by <b>default</b>.</p>
```
Comment (jyn514) [inline templates/crate/features.html:55]:
```suggestion
                    </b> of them enabled by <b>default</b>.</p>
```

File: templates/header/package_navigation.html [hunk 13]
```diff
@@ -85,6 +86,15 @@ <h1 id="crate-title">
                                 <span class="title"> Builds</span>
                             </a>
                         </li>
+
+                        {# The features tab #}
+                        <li class="pure-menu-item">
+                            <a href="/crate/{{ crate_path | safe }}/features"
```
Comment (jyn514) [inline templates/header/package_navigation.html:92]: `| safe` is the default, you don't need to restate it. https://tera.netlify.app/docs/#auto-escaping
```suggestion
                            <a href="/crate/{{ crate_path }}/features"
```

File: templates/rustdoc/topbar.html [hunk 14]
```diff
@@ -229,6 +229,14 @@
                 </li>
             {%- endfor -%}
         </ul>
+    </li>{#
+    Display the features available in current build
+    #}
+    <li class="pure-menu-item">
+        <a href="{{ crate_url | safe }}/features" title="Browse available feature flags of {{ metadata.name }}-{{ metadata.version }}" class="pure-menu-link">
```
Comment (jyn514) [inline templates/rustdoc/topbar.html:236]:
```suggestion
        <a href="{{ crate_url }}/features" title="Browse available feature flags of {{ metadata.name }}-{{ metadata.version }}" class="pure-menu-link">
```

File: src/test/fakes.rs (hunk 11 above)
Comment (almusil) [reply src/test/fakes.rs:218]: They are not optional, package will always have features that are passed from `cargo metadata`: `"features": []`. The option that I have enabled on the last version is that it will show it even for old docs. And feature flags page without features will state it. That seems to be desirable IMO.

File: templates/header/package_navigation.html (hunk 13 above)
Comment (Nemo157) [reply templates/header/package_navigation.html:92]: `safe` declares the string to have already been escaped and _disables_ the auto-escaping.

File: src/test/fakes.rs (hunk 11 above)
Comment (jyn514) [reply src/test/fakes.rs:218]: @almusil what about crates built before this feature was added? Won't it be `NULL` in the database?

File: templates/header/package_navigation.html (hunk 13 above)
Comment (jyn514) [reply templates/header/package_navigation.html:92]: Hmm, ok - @almusil why did you disable the escaping?

Comment (almusil) [reply templates/header/package_navigation.html:92]: I'll be honest, it was copied from the builds link.

File: src/test/fakes.rs (hunk 11 above)
Comment (almusil) [reply src/test/fakes.rs:218]: @jyn514 It will but the frontend part of this is optional. It is kinda confusing. Backend metadata are parsed as non-optional for the reason I have mentioned earlier. Frontend is optional so if the database part is NULL or simply empty it should result in the same thing AFAIK. Maybe I have wrong understanding of the templates and their evaluation of variables.

File: templates/header/package_navigation.html (hunk 13 above)
Comment (jyn514) [reply templates/header/package_navigation.html:92]: Ok, it was probably switched in https://github.com/rust-lang/docs.rs/pull/906 then. No need to change it.

File: src/test/fakes.rs (hunk 11 above)
Comment (jyn514) [reply src/test/fakes.rs:218]: Something weird is going on here, the tests don't match the live version. Notice here the features are optional: https://github.com/rust-lang/docs.rs/blob/2df55efcfde3541e83d0ebe2647a60cb6b266b5c/src/web/mod.rs#L525
but yet you're setting it here to a `HashMap` directly. I don't know why that's possible, but it's not right - we should be testing `None`, not just an empty HashMap.

> Maybe I have wrong understanding of the templates and their evaluation of variables.

It's not _wrong_ but you're not distinguishing between the two I think. `{% if metadata.features %}` is false *either* if the array is empty or if it doesn't exist. It would be nice to say 'no features' if it's empty and 'data not available' if it doesn't exist.

Comment (almusil) [reply src/test/fakes.rs:218]: That's the thing I don't know how to fake the NULL without changing the backend. The change in backend to have option would require some filtering if the features are just empty list from `cargo metadata`.

Comment (jyn514) [reply src/test/fakes.rs:218]: Got it, that makes sense. In that case you can use a direct SQL query - create the package like normal, then `UPDATE releases SET features = NULL WHERE release_id = ?`.

Comment (almusil) [reply src/test/fakes.rs:218]: Ah right, I will add the test and update the frontend to match what you have described NULL vs empty. Thanks

Comment (almusil) [reply src/test/fakes.rs]: Should be done

File: templates/crate/features.html (hunk 12 above)
Comment (almusil) [reply templates/crate/features.html]: Done

File: src/web/crate_details.rs (hunk 10 above)
Comment (almusil) [reply src/web/crate_details.rs]: Done

File: src/web/crate_details.rs [hunk 15]
```diff
@@ -819,4 +816,29 @@ mod tests {
...
     }
+
+    #[test]
+    fn feature_flags_report_null() {
+        wrapper(|env| {
+            let id = env
+                .fake_release()
+                .name("library")
+                .version("0.1.0")
+                .features(HashMap::new())
```
Comment (jyn514) [inline src/web/crate_details.rs:827]: No need to add this call, since you delete the features later.
```suggestion
```

**Review by jyn514 — APPROVED** [review]
This already looks really good and I'm happy to merge it as is :) Some improvements I'd like to see at some point:

- Right now the features look like they're unordered: 
image
Could we change it to put features that enable other features first? Or we could get really fancy and do a tree structure, where we put `default -- [std, perf, unicode]`, then std, then perf, then unicode, then, recursively, features that are enabled by std/perf/unicode.
- We don't distinguish between optional dependencies and features. Could we add that?
- @Nemo157 I think you mentioned a while back you wouldn't want this for your crates - do you still want to add an opt-out to `package.metadata.docs.rs`?
- Could we add some color? Maybe green for `default`, red for features that enable other features, black otherwise, something like that?

**Comment by Nemo157** [comment]
> * 
> 
> * @Nemo157 I think you mentioned a while back you wouldn't want this for your crates - do you still want to add an opt-out to `package.metadata.docs.rs`?

I would, but I wouldn't want to block this on that being added.

I did run across a bug in the feature recording while looking at it though, `async-compression 0.3.6` has a `tokio-02` and a `tokio-03` optional dependency, but these both get recorded and displayed as `tokio`.

File: templates/crate/features.html [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+                    </ul>
+                </div>
+            </div>
+
+            <div class="pure-u-1 pure-u-sm-17-24 pure-u-md-19-24 package-details" id="main">
+                <h1>{{ metadata.name }}</h1>
+                {%- if metadata.features -%}
+                    <p>This version has <b>{{ metadata.features | length }}</b> feature flags, <b data-id="default-feature-len">
+                        {%- if metadata.features[0].name == 'default' -%}
+                            {{ metadata.features[0].subfeatures | length }}
```
Comment (Nemo157) [inline templates/crate/features.html:55]: This count is wrong when one of the default features activates more features. In `bs58 0.4.0` I have `default = ["std"]` and `std = ["alloc"]`.

Also I think the previous count should be decreased by 1 (if the crate has a `default` feature), otherwise a crate which activates all its features by default will still say something like "This version has 5 feature flags, 4 of them enabled by default" because `default` is not counted.

**Comment by almusil** [comment]
@jyn514 @Nemo157 What should be addressed by this PR? The default counting to follow the dependencies?

**Comment by jyn514** [comment]
@almusil if you can fix the bugs Nemo noticed and address my nit I think this is good to merge. The rest can be fixed in follow-ups.

File: templates/crate/features.html (hunk 16 above)
Comment (almusil) [reply templates/crate/features.html:55]: > async-compression 0.3.6

I was trying it locally and I can see both `tokio-02` and `tokio-03`.

Comment (Nemo157) [reply templates/crate/features.html:55]: Since the renamed dependency issue is unrelated to this UI I opened https://github.com/rust-lang/docs.rs/issues/1175 for it.

**Comment by almusil** [comment]
@jyn514 Hopefully got it right this time, thanks for all the patience

File: src/web/features.rs [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+    let mut default_len = 0;
+    let features = row
+        .get::<'_, usize, Option<Vec<Feature>>>(0)
+        .map(|raw| {
+            raw.into_iter()
+                .filter(|feature| !feature.is_private())
+                .map(|feature| (feature.name.clone(), feature))
+                .collect::<HashMap<String, Feature>>()
+        })
+        .map(|mut feature_map| {
```
Comment (jyn514) [inline src/web/features.rs:50]:
```suggestion
            let feature_map: HashMap<_> = raw.into_iter()
                .filter(|feature| !feature.is_private())
                .map(|feature| (feature.name.clone(), feature))
                .collect();
```

File: src/web/features.rs [hunk 18]
```diff
@@ -22,9 +26,63 @@ pub fn build_features_handler(req: &mut Request) -> IronResult<Response> {
...
+            let mut features = get_tree_structure_from_default(&mut feature_map);
+            let mut remaining = feature_map
+                .into_iter()
+                .map(|(_, feature)| feature)
+                .collect::<Vec<Feature>>();
+            remaining.sort_by_key(|feature| feature.subfeatures.len());
```
Comment (jyn514) [inline src/web/features.rs:56]: Can you add some tests for this behavior? Ideally unit tests, but if web tests are easier to write, those are fine too.

File: src/web/features.rs (hunk 17 above)
Comment (almusil) [reply src/web/features.rs]: Done

File: src/web/features.rs (hunk 18 above)
Comment (almusil) [reply src/web/features.rs]: Sure, I have added a few unit tests

File: src/web/features.rs [hunk 19]
```diff
@@ -38,28 +40,14 @@ pub fn build_features_handler(req: &mut Request) -> IronResult<Response> {
...
-            default_len = features.len();
-
-            features.extend(remaining.into_iter().rev());
-            features
-        });
+
+    if let Some(raw) = row.get(0) {
+        let result = order_features_and_count_default_len(raw);
+        features = Some(result.0);
+        default_len = result.1;
```
Comment (jyn514) [inline src/web/features.rs:49]: https://github.com/rust-lang/rust/pull/71156 would help with this :) 'coming soon to rust versions near you'

File: src/web/features.rs [hunk 20]
```diff
@@ -0,0 +1,220 @@
...
+            feature1.clone(),
+        ];
+        let mut feature_map = get_feature_map(raw);
+        let default_tree = get_tree_structure_from_default(&mut feature_map);
+
+        assert_eq!(feature_map.len(), 1);
+        assert_eq!(default_tree.len(), 4);
+        assert!(feature_map.contains_key(&non_default.name));
+        assert!(!feature_map.contains_key(&default.name));
+        assert_eq!(default_tree.get(0), Some(&default));
```
Comment (jyn514) [inline src/web/features.rs:146]: `get` just adds clutter here IMO.
```suggestion
        assert_eq!(default_tree[0], default);
```
and the same for the rest.

File: src/web/features.rs (hunk 19 above)
Comment (almusil) [reply src/web/features.rs:49]: That will be nice :)

File: src/web/features.rs (hunk 20 above)
Comment (almusil) [reply src/web/features.rs]: Done

**Comment by jyn514** [comment]
I think further improvements can be done in follow-ups. Thanks so much for working on this, I've wanted this in docs.rs for a while :heart:

**Comment by almusil** [comment]
Happy to help :) Thank you for your patience. If you don't mind I would like to continue with improvements to this feature.

**Comment by jyn514** [comment]
Sounds great! If you want to talk about ideas we have a [discord channel](https://discord.gg/f7mTXPW) for docs.rs
